bc_obps/Dockerfile

```
@@ -1,58 +1,109 @@
-# Needs to be done as root to chown
-RUN useradd -ms /bin/bash -d ${HOME} --uid ${USER_ID} -g root bc_obps
 
-COPY ./ ${HOME}/
+# Recreate bc_obps user with bash shell
+RUN useradd -ms /bin/bash -d ${HOME} --uid ${USER_ID} -g root bc_obps && \
+    chown -R bc_obps:0 ${HOME} && \
+    chmod -R g+rwX ${HOME}
+
+####################
+# Builder Stage
+####################
+FROM base AS builder
+
+# Refresh package lists from repositories
+RUN apt-get update && \
+# Install minimal build tools required for asdf (git), curl for Poetry plugin, build-essential for compiling Python, and additional libs for Python
+    apt-get install -y --no-install-recommends build-essential curl git libc6-dev libffi-dev libssl-dev zlib1g-dev && \
+# Remove apt cache to reduce image size
+    apt-get clean && \
+# Delete package lists to further slim the image
+    rm -rf /var/lib/apt/lists/*
 
-# Install asdf
-RUN git clone https://github.com/asdf-vm/asdf.git ${HOME}/asdf --depth 1 --branch v0.11.2
+# Copy version config and Poetry files with ownership set to bc_obps
+COPY --chown=bc_obps:0 .tool-versions pyproject.toml poetry.lock ${HOME}/
 
+USER ${USER_ID}
+
+# Clone asdf shallowly from GitHub
+RUN git clone --depth 1 --branch v0.11.2 https://github.com/asdf-vm/asdf.git ${HOME}/asdf
+
+# Ensure asdf is available in non-interactive shells
 ENV BASH_ENV="${HOME}/asdf/asdf.sh"
-# Because asdf is loaded via BASH_ENV, all commands using adsf need to be executed using /usr/bin/env bash -c
-SHELL ["/usr/bin/env", "bash", "-c"]
 
+# Use bash with environment sourcing for asdf commands
+SHELL ["/bin/bash", "-c"]
+
+# Filter .tool-versions for python and poetry
 RUN sed -i -nr '/python|poetry/p' ${HOME}/.tool-versions && \
+# Add plugins for each tool listed in .tool-versions
     cat ${HOME}/.tool-versions | cut -f 1 -d ' ' | xargs -n 1 asdf plugin-add && \
+# Update all plugins to latest versions
     asdf plugin-update --all && \
+# Install tools (Python, Poetry) from .tool-versions
     asdf install && \
+# Regenerate shims for installed tools
     asdf reshim
 
-# Extract the Poetry version dynamically
-ARG POETRY_VERSION
+# Extract Poetry version from .tool-versions
 RUN export POETRY_VERSION=$(grep '^poetry ' ${HOME}/.tool-versions | awk '{print $2}') && \
-    echo "Using Poetry version: $POETRY_VERSION"
-
-# Set the Poetry version as an environment variable
-ENV POETRY_VERSION=${POETRY_VERSION}
-ENV PATH="${HOME}/.asdf/installs/poetry/${POETRY_VERSION}/bin:${PATH}"
+# Add Poetry to PATH in bashrc
+    echo "export PATH=${HOME}/.asdf/installs/poetry/${POETRY_VERSION}/bin:${PATH}" >> ${HOME}/.bashrc
 
 RUN poetry config virtualenvs.create true && \
-	poetry config virtualenvs.in-project true
+# Keep virtualenv in project directory (.venv)
+    poetry config virtualenvs.in-project true && \
+# Install dependencies without dev group, non-interactively
+    poetry install --without dev --no-root --no-interaction --no-ansi
```

Let's build the bytecode too!
```suggestion
    poetry install --without dev --no-root --no-interaction --no-ansi --compile
```

---

In the same vein,  do you think we'd gain at doing `poetry build` here too?

---

I’m not sure if this is applicable for building a web app. I mostly used it for building packages.